crates/accelerate/src/nlayout.rs

```diff
@@ -72,6 +73,25 @@ impl PhysicalQubit {
         layout.phys_to_virt[self.index()]
     }
 }
+
+/// This is only safe in the context of a physical circuit during transpilation
+/// after the qubit indices of the circuit/dag circuit refer to the physical
+/// qubits (once we've run a layout pass and applied it).
+impl From<Qubit> for PhysicalQubit {
+    fn from(s: Qubit) -> PhysicalQubit {
+        PhysicalQubit::new(s.0)
+    }
+}
+
+/// This is only safe in the context of a physical circuit during transpilation
+/// after the qubit indices of the circuit/dag circuit refer to the physical
+/// qubits (once we've run a layout pass and applied it).
+impl From<&Qubit> for PhysicalQubit {
+    fn from(s: &Qubit) -> PhysicalQubit {
+        PhysicalQubit::new(s.0)
+    }
+}
+
```

Comment:
These are pretty huge caveats on a really core trait implementation.  I think leaving it as requiring `PhysicalQubit::new(q.0)` is probably better, or adding `Qubit::as_physical` or something, so you always have to be explicit about what you're doing.  `.into()`'s a bit magical to have any sort of caveat for me, personally.

Reply:
That's a fair point, I'll add a dedicated method to `Qubit` for it. I started this by doing `PhysicalQubit::new(q.0)` and thought it was a bit tedious to do it that way.

Reply:
Actually I took a quick look at doing this and it will require some reorganization of the code I think. The problem is `PhysicalQubit` is defined in `accelerate::nlayout` and `Qubit` is defined in `qiskit_circuit`. I don't want to add a bidirectional dependency between `circuit` and `accelerate`. I think in a separate PR we'll want to probably move `nlayout` into circuit (as `PhysicalQubit` and `VirtualQubit`'s methods have a dependency on `NLayout`) and define all the `Qubit`  types together. Then we can do `Qubit::as_physical` as part of that reorganization.

Reply:
I removed the trait impl for this here: https://github.com/Qiskit/qiskit/pull/12959/commits/9b43ea7c02fdba8d8c4ea6a4c70c5961f4faf61a